src/DataStructures/DataBox/DataBox.hpp

```diff
@@ -1082,36 +1082,35 @@ SPECTRE_ALWAYS_INLINE constexpr void mutate_apply(
       "Cannot pass a DataBox to mutate_apply since the db::get won't work "
       "inside mutate_apply.");
   if constexpr (is_apply_callable_v<
-                    F, const gsl::not_null<item_type<ReturnTags, BoxTags>*>...,
```

Comment:
(picking a random line) Question: Should we eventually remove the ability to call `db::apply` and friends with a "stateless invocable" (i.e. a class with an `apply` function) from which we infer the `argument_tags`, and just require the invocable is callable and the user provides the argument tags explicitly? That would simplify this logic a lot (e.g. removes the need to check `tt::is_callable` since it just tries to call the invocable and lets the compiler figure out if it works). It would also bring this in line with `tuples::apply` and make it resemble `std::apply` more closely. What do you think?

Reply:
to clarify: you are proposing renaming the `apply` functions in the `stateless invocables` to `operator()`?

Reply:
yes, and require the `argument_tags` list is always passed to `db::apply` explicitly, instead of sometimes inferring it from the invocable.

Reply:
but then we'd have to construct the object, which is exactly what we are trying at avoid. Maybe in C++17 you can have `static operator()` but you couldn't before, I think

Reply:
okay sure I am fine with the renaming of `apply` to `operator()` to simplify the type logic; whether or not you specify the argument_tags directly or through a type alias  is handled in the simple public functions which call the more complicated detail function

Reply:
oh right forgot about constructing the object...

Reply:
1. Are compilers smart enough to optimize away the overhead from constructing the object?
2. Do we use these "stateless invocables" anywhere where we couldn't just replace them with lambda functions? If I remember correctly the CCE code does, actually.
3. The rename `apply` -> `operator()` would make the `db::apply` logic a lot simpler, thus potentially reducing compile time at the cost of potentially a minor runtime-overhead from constructing the object. I have no idea how relevant either of these costs actually are 🤷‍♂️
4. Requiring the `argument_tags` are explicitly specified seems independent of the `apply`/`operator()` issue. What do you think about that?

Reply:
wait, the stateless object is already constructed (either to be passed in to the function directly or specified via template parameter, in which case it is default constructed and passed in to the function...)

Reply:
Oh right...

Reply:
1. can't really count on that. Constructing an object can have a lot of side effects, and you _do_ use the object, so I see no way that the compiler could remove the construction.
2. I guess you can technically wrap every call in a closure, but that seems to be a bit going backwards. Now every time you want to call an `apply` function you need to write a wrapper closure.
3. Construction overhead could be extremely high. Just blindly executing arbitrary code is always bad idea. The constructor could do a lot of work, work that might not be needed for the call even.
4. We used to have that restriction, and I think you wanted the ability not to have to do `db::apply<T::argument_list>(t);` or whatever the notation is. I'm fine always having to specify where the tags are coming from, but probably not in this PR.